File: flower/src/main/java/com/kharitonov/flower/controllers/BasketController.java
package com.kharitonov.flower.controllers;

import com.kharitonov.flower.contracts.ApiRouter;
import com.kharitonov.flower.dtos.BasketDTO;
import com.kharitonov.flower.entities.BasketEntity;
import com.kharitonov.flower.entities.FlowerEntity;
import com.kharitonov.flower.services.BasketService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Optional;

@RestController
@RequestMapping(ApiRouter.BasketFlower.BASE_URL)
public class BasketController {

    private final BasketService basketService;

    @Autowired
    public BasketController(BasketService basketService) {
        this.basketService = basketService;
    }

    @PreAuthorize("hasAuthority('USER')")
    @GetMapping
    public ResponseEntity<List<BasketDTO>> findAll() {
        List<BasketDTO> basketDTOS = basketService.findAll();
        return ResponseEntity.ok(basketDTOS);
    }

    @PreAuthorize("hasAuthority('USER')")
    @GetMapping(ApiRouter.BasketFlower.FIND_BY_ID)
    public BasketDTO findById(@PathVariable("id") Long id) {
        return basketService.findById(id);
    }

    @PreAuthorize("hasAuthority('USER')")
    @PostMapping
    public ResponseEntity<BasketDTO> create(@RequestBody Optional<BasketEntity> basketEntity) {
        BasketDTO basketDTO = basketService.save(basketEntity);
        return ResponseEntity.ok(basketDTO);
    }

    @PreAuthorize("hasAuthority('USER')")
    @PutMapping(ApiRouter.BasketFlower.UPDATE)
    public ResponseEntity<BasketDTO> update(@PathVariable("id") Optional<BasketEntity> basketEntityFromDB,
                            @RequestBody Optional<BasketEntity> basketEntity) {
        BasketDTO basketDTO = basketService.update(basketEntityFromDB, basketEntity);
        return ResponseEntity.ok(basketDTO);
    }

    @PreAuthorize("hasAuthority('USER')")
    @PutMapping(ApiRouter.BasketFlower.ADD_FLOWER)
    public ResponseEntity<BasketDTO> addFlower(@PathVariable("toBasketId") Optional<BasketEntity> basketEntity,
                               @PathVariable("flowerId")Optional<FlowerEntity> flowerEntity) {
        BasketDTO basketDTO = basketService.addFlower(basketEntity, flowerEntity);
        return ResponseEntity.ok(basketDTO);
    }

    @PreAuthorize("hasAuthority('USER')")
    @DeleteMapping(ApiRouter.BasketFlower.DELETE_BY_ID)
    public void delete(@PathVariable("id") Optional<BasketEntity> basketEntity) {
        basketService.delete(basketEntity);
    }


}

File: src/main/java/by/hghgrtut/shop/controllers/BasketController.kt
package by.hghgrtut.shop.controllers

import by.hghgrtut.shop.contracts.ApiRouter
import by.hghgrtut.shop.model.BasketModel
import by.hghgrtut.shop.entities.BasketEntity
import by.hghgrtut.shop.entities.ItemEntity
import by.hghgrtut.shop.services.BasketService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.ResponseEntity
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping(ApiRouter.BasketItem.BASE_URL)
class BasketController @Autowired constructor(private val basketService: BasketService) {

    @PreAuthorize(PREAUTHORIZE_AUTHORITY)
    @GetMapping
    fun findAll(): ResponseEntity<List<BasketModel>> = getEntity(basketService.findAll())

    @PreAuthorize(PREAUTHORIZE_AUTHORITY)
    @GetMapping(ApiRouter.BasketItem.FIND_BY_ID)
    fun findById(@PathVariable("id") id: Long?): BasketModel = basketService.findById(id)

    @PreAuthorize(PREAUTHORIZE_AUTHORITY)
    @PostMapping
    fun create(@RequestBody basketEntity: BasketEntity?): ResponseEntity<BasketModel> =
            getEntity(basketService.save(basketEntity))

    @PreAuthorize(PREAUTHORIZE_AUTHORITY)
    @PutMapping(ApiRouter.BasketItem.UPDATE)
    fun update(@PathVariable("id") basketEntityFromDB: BasketEntity?, @RequestBody basketEntity: BasketEntity?) =
            getEntity(basketService.update(basketEntityFromDB, basketEntity))

    @PreAuthorize(PREAUTHORIZE_AUTHORITY)
    @PutMapping(ApiRouter.BasketItem.ADD_FLOWER)
    fun addItem(
            @PathVariable("toBasketId") basketEntity: BasketEntity?,
            @PathVariable("itemId") itemEntity: ItemEntity?
    ): ResponseEntity<BasketModel> = getEntity(basketService.addItem(basketEntity, itemEntity))

    @PreAuthorize(PREAUTHORIZE_AUTHORITY)
    @DeleteMapping(ApiRouter.BasketItem.DELETE_BY_ID)
    fun delete(@PathVariable("id") basketEntity: BasketEntity?) { basketService.delete(basketEntity) }
    
    private fun getEntity(basketModel: BasketModel) = ResponseEntity.ok(basketModel)
}